### ISSUE TITLE
Mimimal template for the APM indexing strategy

### ISSUE THREAD

**Issue by jalvz** [elastic/elasticsearch#64047]
For the upcoming APM integration we need a new `traces` data type, as described in the [Indexing Strategy proposal](https://docs.google.com/document/d/1y56a9fjkLi6Zen5qGC_JKYM9ljYpBA5W0fgdWilcwYc/edit?ts=5f6c1aab#).

This in turn requires an index template bundled by default in Elasticsearch matching the index pattern `traces-*-*`.

[EPR Issue](https://github.com/elastic/package-spec/issues/68).

Related: https://github.com/elastic/elasticsearch/issues/56709

**Comment by dakrone** [comment]
@jalvz @ruflin can we dig into this a bit more? My understanding for Elasticsearch shipping templates rather than relying on the fleet/integrations templates was that in some cases users would use Elasticsearch separately from the ingest manager. I think for logs, metrics, and synthetics this makes sense. For APM traces however, I'm not sure I currently see the use case where a user will set up APM agents without any integrations through Fleet. Can someone explain why this may or may not be the case?

**Comment by ruflin** [comment]
My assumption so far is that there are users that will run Agent with APM Server also in Standalone mode potentially before going to Kibana. But that assumption might be wrong. If users will always go to Kibana first, agree the `traces` part is not needed. One note here: The Agent / APM Server will not know if the setup was done or not so we might still need the `traces` template a measure to make sure not wrong automatic setup happens.

@jalvz @axw @graphaelli  Can you comment on the question from @dakrone ?

**Comment by jalvz** [comment]
I can't see any use for standalone, my understanding was that we needed it because of this (sort of): 

> The Agent / APM Server will not know if the setup was done or not so we might still need the traces template a measure to make sure not wrong automatic setup happens

I can imagine a user copy-pasting an APM policy snippet (instead of going trough the UI), indexing some data, realize that nothing works, come back to the UI, install the APM integration: 
Does it makes a difference whether there was (previous any ingestion) a template for traces with data streams enabled or not? 
In other words: I was under the impression that in the accidental case of ingest-before-install, without a data stream template users would have to delete their indices, otherwise they would not. 

If this is not the case, or if it this is not an argument strong enough, then I'm happy to just close the issue.

**Comment by ruflin** [comment]
@jalvz Yes, in the accidental case users would have to delete the data because it would create indices instead of data streams. If we have the template, it is all data streams and just a rollover happens to use the more specific mappings.

**Comment by jalvz** [comment]
Gotcha. Then I assume that in case of an accident, users will notice rather quickly, and having to delete data is not all that bad. 

Still it would be better UX to not have to IMO (just install the integration and continue like nothing happened), so from my perspective a default template for traces is a nice to have and not a must.

**Comment by ruflin** [comment]
How will the user know which data to delete? Will we provide him some tooling to clean up? I personally think it is a less then ideal user experience especially because we could prevent it form happening. Worst case is if a user starts to use `traces-*-*` for something else so he can't delete the data but also not use APM ...

**Comment by jalvz** [comment]
> Will we provide him some tooling to clean up?

I don't think so

> How will the user know which data to delete?

Docs? :grimacing: 

But I agree with you, not great UX. I don't think that the worse case you describe is very likely (albeit certainly possible), but I also think that this kind of accident can happen fairly often. 

Personally I am favourable to move forward with this, but I don't know either the drawbacks of bundling another template in Elasticsearch.

**Comment by simitt** [comment]
If we ignore the standalone case for a moment (which I believe we don't need to optimize for in APM), can we build the fleet - agent input creation process in a way that ensures the templates in ES are set up before the APM Server is actually started? This way we wouldn't need any shipped templates.

**Comment by dakrone** [comment]
> can we build the fleet - agent input creation process in a way that ensures the templates in ES are set up before the APM Server is actually started? This way we wouldn't need any shipped templates.

This would be my preference, because it's more correct in the event that any ES template installation is disabled.

**Comment by jalvz** [comment]
I would prefer it to, so far I've been assuming it is not possible - it would require a fairly major re-achitecture of Elastic Agent to deal with packages, which I believe is intentionally out of scope (@ruflin correct me if wrong).

However, I just learned about https://github.com/elastic/elasticsearch/pull/61525

IIUC that means we can migrate legacy indices to data streams without data loss?

**Comment by dakrone** [comment]
> IIUC that means we can migrate legacy indices to data streams without data loss?

That is only possible if an alias is set up, which wouldn't be if the template were not installed.

I guess I'm a little lost on how Elastic Agent deals with other packages that don't have their infrastructure correctly set up prior to the agent running, wouldn't this become a major issue as the barebones ES templates diverge from the Fleet ones? Shouldn't Fleet always ensure that the scaffolding is set up correctly prior to starting the agent?

**Comment by axw** [comment]
> If we ignore the standalone case for a moment (which I believe we don't need to optimize for in APM), can we build the fleet - agent input creation process in a way that ensures the templates in ES are set up before the APM Server is actually started? This way we wouldn't need any shipped templates.

This is how things should work when running APM Server under Elastic Agent in Fleet mode. This will be the default for APM Server.

The problems described above may occur when running APM Server under Elastic Agent in standalone mode. If users do this they'll be expected to perform more manual setup.

It seems to me that the primary (if not only) benefit of providing a base template would be to ensure a data stream is created in all cases, instead of a plain old index, and only under exceptional circumstances (i.e. in standalone mode). It feels wrong to me to have a template for only this.

As an alternative, could we limit the damage done in this scenario by preventing plain old indices from being created? Remediation then involves only setting up the correct templates, and nothing needs to be deleted. A couple of options come to mind:
 - provide an option for Elasticsearch clients to indicate that a document should only be indexed if the target is a data stream (existing or auto created). Somewhat like the `require_alias` query parameter option.
 - provide APM Server with a privilege narrower than `indices:admin/auto_create`, which can only auto create data streams.

As the base template is unnecessary for standard operation, and it's unclear how big of a problem this failure mode will be, I suggest we close the issue and revisit our options once we're a bit further along.

**Comment by jalvz** [comment]
Closing then, thanks all.

**Comment by ruflin** [comment]
@axw I like your idea around having only permissions to create data-streams and not indices. Should we follow up on this in a separate issue?

**Comment by axw** [comment]
> I like your idea around having only permissions to create data-streams and not indices. Should we follow up on this in a separate issue?

I'm unsure if we'll really need this functionality, so unless others want it I'll wait before opening a feature request.
